### DISCUSSION TITLE
How evaluation metrics are calculated?

### DISCUSSION THREAD

**Discussion by KennyTC** [talmolab/sleap#1045] [Help!]
After running training models, I see a lot of evaluation metrics, for example

vis.tp
vis.fp
vis.tn
vis.fn
vis.precision
vis.recall
dist.dists
dist.avg
dist.p50
dist.p75
dist.p90
dist.p95
dist.p99
pck.thresholds
pck.pcks
pck.mPCK_parts
pck.mPCK
oks.mOKS
oks_voc.match_score_thresholds
oks_voc.recall_thresholds
oks_voc.match_scores
oks_voc.precisions
oks_voc.recalls
oks_voc.AP
oks_voc.AR
oks_voc.mAP
oks_voc.mAR
pck_voc.match_score_thresholds
pck_voc.recall_thresholds
pck_voc.match_scores
pck_voc.precisions
pck_voc.recalls
pck_voc.AP
pck_voc.AR
pck_voc.mAP

For distance related metrics such as (dist.dists, dist.avg, dist.p50, dist.p75.....), I can understand. 
But for other metrics, I could not find how these metrics are calculated. How can I get the overview of how these metrics calculated? 

Any recommendation of which metrics should I focus when comparing these models?

**Comment by roomrys** [comment]
Hi @KennyTC,

I'll start by listing the names of the metrics below:
        - `"vis.tp"`: Visibility - True Positives
        - `"vis.fp"`: Visibility - False Positives
        - `"vis.tn"`: Visibility - True Negatives
        - `"vis.fn"`: Visibility - False Negatives
        - `"vis.precision"`: Visibility - Precision
        - `"vis.recall"`: Visibility - Recall
        - `"dist.avg"`: Average Distance (ground truth vs prediction)
        - `"dist.p50"`: Distance for 50th percentile
        - `"dist.p75"`: Distance for 75th percentile
        - `"dist.p90"`: Distance for 90th percentile
        - `"dist.p95"`: Distance for 95th percentile
        - `"dist.p99"`: Distance for 99th percentile
        - `"dist.dists"`: All distances
        - `"pck.mPCK"`: Mean Percentage of Correct Keypoints (PCK)
        - `"oks.mOKS"`: Mean Object Keypoint Similarity (OKS)
        - `"oks_voc.mAP"`: VOC with OKS scores - mean Average Precision (mAP)
        - `"oks_voc.mAR"`: VOC with OKS scores - mean Average Recall (mAR)
        - `"pck_voc.mAP"`: VOC with PCK scores - mean Average Precision (mAP)
        - `"pck_voc.mAR"`: VOC with PCK scores - mean Average Recall (mAR)

Calculations are tied together through the [`evaluate`](https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L585-L647) function.

## All visibility (vis):
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L549-L582

## The distances (dist.dists):
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L472-L492

## The distance errors (dist.avg, dist.pXX):
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L495-L520

## All Percentage of Correct Points (pck):
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L523-L546

## Mean Object Keypoint Similarity (oks.mOKS)
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L632-L635
where `positive_pairs` is calculated here
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L340-L386

## All Visual Object Classes (_voc):
https://github.com/talmolab/sleap/blob/5093f6992e6214c0d528b7240331b99d0a89a62f/sleap/nn/evals.py#L389-L469

The [model evaluation notebook](https://sleap.ai/notebooks/Model_evaluation.html?highlight=model) provides further info on what metrics might be helpful to look at.

Hope this helps!

Thanks,
Liezl
